pkg/pillar/cmd/zedagent/handlecipherconfig.go

```diff
@@ -133,134 +59,196 @@ func parseCipherContextConfig(getconfigCtx *getconfigContext,
 			log.Debugf("parseCipherContextConfig ignoring empty\n")
 			continue
 		}
-		cipherContext := new(types.CipherContext)
-		cipherContext.ID = cfgCipherContext.GetContextId()
-		cipherContext.HashScheme = cfgCipherContext.GetHashScheme()
-		cipherContext.KeyExchangeScheme = cfgCipherContext.GetKeyExchangeScheme()
-		cipherContext.EncryptionScheme = cfgCipherContext.GetEncryptionScheme()
-		cipherContext.DeviceCertHash = cfgCipherContext.GetDeviceCertHash()
-		cipherContext.ControllerCertHash = cfgCipherContext.GetControllerCertHash()
-		log.Debugf("parseCipherContextConfig publishing %v\n", cipherContext)
-		publishCipherContextConfig(getconfigCtx, cipherContext)
+		cipherCtx := types.CipherContext{}
+		cipherCtx.ContextID = cfgCipherContext.GetContextId()
+		cipherCtx.HashScheme = cfgCipherContext.GetHashScheme()
+		cipherCtx.KeyExchangeScheme = cfgCipherContext.GetKeyExchangeScheme()
+		cipherCtx.EncryptionScheme = cfgCipherContext.GetEncryptionScheme()
+		cipherCtx.DeviceCertHash = cfgCipherContext.GetDeviceCertHash()
+		cipherCtx.ControllerCertHash = cfgCipherContext.GetControllerCertHash()
+		updateCipherContextCerts(ctx, &cipherCtx)
+		publishCipherContextConfig(ctx, cipherCtx)
 	}
 }
 
-func publishCipherContextConfig(getconfigCtx *getconfigContext,
-	config *types.CipherContext) {
+// on create/modification, update the cipher context certs
+func updateCipherContextCerts(ctx *getconfigContext,
+	cipherCtx *types.CipherContext) {
+	// get controller cert
+	if len(cipherCtx.ControllerCert) == 0 {
+		cert := getControllerCert(ctx.zedagentCtx,
+			cipherCtx.ControllerCertHash)
+		cipherCtx.ControllerCert = cert
+	}
 
-	key := config.Key()
-	log.Debugf("publishCipherContext %s\n", key)
-	pub := getconfigCtx.pubCipherContextConfig
-	pub.Publish(key, *config)
+	// get device cert
+	if len(cipherCtx.DeviceCert) == 0 {
+		cert := getDeviceCert(cipherCtx.HashScheme,
+			cipherCtx.DeviceCertHash)
+		cipherCtx.DeviceCert = cert
+	}
 }
 
-func unpublishCipherContextConfig(getconfigCtx *getconfigContext, key string) {
+// parseCipherBlock : will collate all the relevant information
+// ciphercontext will be used to get the certs and encryption schemes
+func parseCipherBlock(ctx *getconfigContext,
+	cfgCipherBlock *zconfig.CipherBlock) types.CipherBlock {
+	cipherBlock := types.CipherBlock{}
+	if cfgCipherBlock == nil {
+		return cipherBlock
+	}
```

Comment:
Can't be nil; you're creating a struct.

Reply:
The cfgCipherBlock is coming from controller, as nil. The CipherBlock returns a structure.